### ISSUE TITLE
AWS Sigv4 incorrect when '/' in file: ?

### ISSUE THREAD

**Issue by woooooopwop** [TTLabs/EvaporateJS#196]
Hi,

I used to upload file to S3 under a specific prefix using the following code (with AWS sigv2)

```javascript
evaporate.add({
                    name: userEmailAddress + '/' + projectName + '/'+ sanitizedName,
                    ...
});
```

I have now moved to AWS v4 signature, and noticed my code did not work anymore.
I tried removing the '/', and I was able to upload the file successfully. 

The canonical request seems to be generated correctly, as a trailing slash is added at the beginning of my request as expected

```
/bla%40email.com%2Fbla%2FScreen%20Shot%202016-08-08%20at%2010.17.31%20AM.png
uploads=
host:mybucket.s3-accelerate.amazonaws.com
x-amz-date:20160810T175647Z
```

Not sure what's wrong here. Is there any other way to specify a specific prefix to a S3 upload when you use v4?

**Comment by bikeath1337** [comment]
AWS should respond with useful error messages in XML. Please provide that XML for the failing call. You should be able to copy the response from your browser's developer tools page related to XHR resources/requests.

Please also provide sample code of how you are instantiating Evaporate.

I see you're using transfer acceleration. If this wasn't part of your V2 signature configuration, I would check that out, too. Thanks!

**Comment by woooooopwop** [comment]
Thanks for looking into it.
Transfer acceleration was working fine with V2, and works fine with V4 when the file name does not contain any '/', so I think we can safely assume the problem is not with the S3 endpoint

evaporate.js
```javascript
 // to enable Evaporate
        var _e_ = new Evaporate({
            signerUrl: '/sign_auth4',
            aws_key: 'MYKEY',
            awsRegion: 'ap-southeast-1',
            bucket: 'MYBUCKET',
            aws_url: 'https://s3-ap-southeast-1.amazonaws.com',
            awsSignatureVersion: '4',
            computeContentMd5: true,
            cryptoMd5Method: function (data) { return AWS.util.crypto.md5(data, 'base64'); },
            cryptoHexEncodedHash256: function (data) { return AWS.util.crypto.sha256(data, 'hex'); },
            logging: true,
            maxFileSize: 'no',
            partSize: 6 * 1024 * 1024,
            cloudfront: true,
            s3Acceleration: true

        });
```

My back-end 
```python
def sign(key, msg):
    return hmac.new(key, msg.encode('utf-8'), hashlib.sha256).digest()


def getSignatureKey():
    t = datetime.datetime.utcnow()
    datestamp = t.strftime('%Y%m%d')
    kDate = sign('AWS4MYAWSSECUREKEY'.encode('utf-8'), datestamp)
    kRegion = sign(kDate, 'ap-southeast-1')
    kService = sign(kRegion, 's3')
    kSigning = sign(kService, 'aws4_request')
    return kSigning


def sign_auth4(request):
    to_sign = request.GET['to_sign']
    signing_key = getSignatureKey()
    signature = hmac.new(signing_key, (to_sign).encode('utf-8'), hashlib.sha256).hexdigest()
    print signature
    return HttpResponse(signature)
```

Working example:
```javascript
 _e_.add({
                    name: file['name'],
                    file: file,
```

Canonical request:
```
CanonicalRequest POST
/Screen%20Shot%202016-08-08%20at%2010.17.31%20AM.png
uploadId=5S7r7Jab52GzVtkSRhnKDHys8mdV177e.mB7PVOkABHIs9n2bnB2GFRIqzYCJGvrbebQ6osVfqXn7SqCIyKyt.EKIF_qd6FnVIHYzKO6zDKQxx.bgp7Z_OD2b7QlOXHoPEV2Ng5ZLr7Kq2odflV8Dg--
content-type:application/xml; charset=UTF-8
host:MYBUCKET.s3-accelerate.amazonaws.com
x-amz-date:20160810T183910Z
```

Non-working example:
```javascript
_e_.add({
                    name: userEmailAddress + '/' + projectName + '/'+ file['name'],
                    file: file,
```

Canonical request:
```
CanonicalRequest POST
/user%40email.com%2Fproject%2FScreen%20Shot%202016-08-08%20at%2010.17.31%20AM.png
uploads=
host:MYBUCKET.s3-accelerate.amazonaws.com
x-amz-date:20160810T184413Z

```

XML response (403)

```
<Error>
<Code>SignatureDoesNotMatch</Code>
<Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message>
<AWSAccessKeyId>MYKEY</AWSAccessKeyId>
<StringToSign>AWS4-HMAC-SHA256
20160810T184423Z
20160810/ap-southeast-1/s3/aws4_request
95a99edac9237eba9cb493d4093672ee02fe4a623857a162f84e51a1427ec4af</StringToSign>
<SignatureProvided>fe3254d8b18cba8ae937ca01025fe5582e1e9bba893937e7a40be232c7048b2f</SignatureProvided>
<StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 31 36 30 38 31 30 54 31 38 34 34 32 33 5a 0a 32 30 31 36 30 38 31 30 2f 61 70 2d 73 6f 75 74 68 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 39 35 61 39 39 65 64 61 63 39 32 33 37 65 62 61 39 63 62 34 39 33 64 34 30 39 33 36 37 32 65 65 30 32 66 65 34 61 36 32 33 38 35 37 61 31 36 32 66 38 34 65 35 31 61 31 34 32 37 65 63 34 61 66</StringToSignBytes>
<CanonicalRequest>POST
/user%40email.com/project/Screen%20Shot%202016-08-08%20at%2010.17.31%20AM.png
uploads=
host:MYBUCKET.s3-accelerate.amazonaws.com
x-amz-date:20160810T184423Z

host;x-amz-date
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855</CanonicalRequest>
<CanonicalRequestBytes>50 4f 53 54 0a 2f 6d 63 72 6f 7a 65 73 25 34 30 6c 61 62 31 32 36 2e 63 6f 6d 2f 53 6c 6f 61 6e 65 2f 53 63 72 65 65 6e 25 32 30 53 68 6f 74 25 32 30 32 30 31 36 2d 30 38 2d 30 38 25 32 30 61 74 25 32 30 31 30 2e 31 37 2e 33 31 25 32 30 41 4d 2e 70 6e 67 0a 75 70 6c 6f 61 64 73 3d 0a 68 6f 73 74 3a 77 69 6e 64 63 68 69 6c 6c 73 69 6d 70 6c 65 64 6f 63 75 6d 65 6e 74 73 68 61 72 69 6e 67 2e 73 33 2d 61 63 63 65 6c 65 72 61 74 65 2e 61 6d 61 7a 6f 6e 61 77 73 2e 63 6f 6d 0a 78 2d 61 6d 7a 2d 64 61 74 65 3a 32 30 31 36 30 38 31 30 54 31 38 34 34 32 33 5a 0a 0a 68 6f 73 74 3b 78 2d 61 6d 7a 2d 64 61 74 65 0a 65 33 62 30 63 34 34 32 39 38 66 63 31 63 31 34 39 61 66 62 66 34 63 38 39 39 36 66 62 39 32 34 32 37 61 65 34 31 65 34 36 34 39 62 39 33 34 63 61 34 39 35 39 39 31 62 37 38 35 32 62 38 35 35</CanonicalRequestBytes>
<RequestId>F2677CD31E1B4300</RequestId>
<HostId>c5dbBko4WNw8RPDHmJtS+vXHx7q+q9eF+ALAykenNaBep1rl4k9PElE88haOd56/7W+Pqz3f5bU=</HostId>
</Error>
```
Looks like AWS see the correct path (<CanonicalRequest>POST
/user%40email.com/Project/Screen%20Shot%202016-08-08%20at%2010.17.31%20AM.png) but for some reason the signature does not seems to be valid

**Comment by bikeath1337** [comment]
Thanks, I think it has to do with escaping...I can reproduce here and am looking at it now.

**Comment by woooooopwop** [comment]
You guessed it right, I replaced with any other char and it works. I will try to make it work but my js skills are limited =P

**Comment by bikeath1337** [comment]
I found the problem and will fix it shortly.

**Comment by bikeath1337** [comment]
Fixed in v1.5.1. Thanks for the report. The V4 signature is newish and I'm glad to work out the remaining kinks.

**Comment by woooooopwop** [comment]
Thanks !
